### PR TITLE
REF: consolidate DTI/TDI/PI get_value in ExtensionIndex, associated cleanups

### PR DESCRIPTION
**Pull request by jbrockmendel**
@jschendel does the holds_integer change for IntervalIndex make sense?

@jreback is this a correct usage of self.holds_integer() as meaning "we only fall back to positional indexing if not self.holds_integer()"

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -526,6 +522,10 @@ def is_overlapping(self) -> bool:
         # GH 23309
         return self._engine.is_overlapping
 
+    def holds_integer(self):
+        return self.dtype.subtype.kind not in ["m", "M"]
```
Comment (jreback) [inline pandas/core/indexes/interval.py]: this is ok, but doesn't follow the other patterns of inference, where we use infer_dtype. but for known dtypes this is (sligthly) cheaper.

Comment (jreback) [reply pandas/core/indexes/interval.py]: see holds_integer in the indexes/base.py

Comment (jbrockmendel) [reply pandas/core/indexes/interval.py]: @jschendel the Float64Index behavior was changed in #31329 so s_fi.at[1] returns "b".  This was done to make .at behavior match .loc behavior.

Deferring to self.left or self.right seems reasonable, but we may need something other than holds_integer.  Most of the places where it is used are de-facto used like it is here, but that doesn't match the docstring description.

**Review by jschendel — APPROVED** [review]

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (jreback) [reply pandas/core/indexes/interval.py]: this seems fine for now, @jbrockmendel if you can make a note to re-visit later

**Comment by jreback** [comment]
can you rebase

**Comment by jbrockmendel** [comment]
rebased+green

**Comment by jreback** [comment]
thanks

**Pull request merged**